src/zkevm_specs/state.py

```diff
@@ -0,0 +1,540 @@
+    value: FQ
+    auxs: Tuple[FQ, FQ]
+    # fmt: on
+
+    def tag(self):
+        return self.keys[0]
+
+
+def linear_combine(limbs: Sequence[FQ], base: FQ) -> FQ:
+    ret = FQ.zero()
+    for limb in reversed(limbs):
+        ret = ret * base + limb
+    return ret
+
+
+# Boolean Expression builder
+def all_keys_eq(row: Row, row_prev: Row) -> bool:
+    eq = True
+    for i in range(len(row.keys)):
+        eq = eq and (row.keys[i] == row_prev.keys[i])
+    return eq
+
+
+@is_circuit_code
+def assert_in_range(x: FQ, min_val: int, max_val: int):
+    assert min_val <= x.n and x.n <= max_val
+
+
+@is_circuit_code
+def check_memory(row: Row, row_prev: Row):
+    get_call_id = lambda row: row.keys[1]
+    get_mem_addr = lambda row: row.keys[2]
+
+    # 0. Unused keys are 0
+    assert row.keys[2] == 0
+    assert row.keys[4] == 0
+
+    # 1. First access for a set of all keys
+    #
+    # When the set of all keys changes (first access of an address in a call)
+    # - If READ, value must be 0
+    if not all_keys_eq(row, row_prev) and row.is_write == 0:
+        assert row.value == 0
+
+    # 2. mem_addr in range
+    mem_addr = get_mem_addr(row)
+    assert_in_range(mem_addr, 0, MAX_MEMORY_ADDRESS)
+
+    # 3. value is a byte
+    assert_in_range(row.value, 0, 2**8 - 1)
+
+
+@is_circuit_code
+def check_stack(row: Row, row_prev: Row):
+    get_call_id = lambda row: row.keys[1]
+    get_stack_ptr = lambda row: row.keys[3]
+
+    # 0. Unused keys are 0
+    assert row.keys[2] == 0
+    assert row.keys[4] == 0
+
+    # 1. First access for a set of all keys
+    #
+    # The first stack operation in a stack position is always a write (can't
+    # read if it isn't written before)
+    #
+    # When the set of all keys changes (first access of a stack position in a call)
+    # - It must be a WRITE
+    if not all_keys_eq(row, row_prev):
+        assert row.is_write == 1
+
+    # 2. stack_ptr in range
+    stack_ptr = get_stack_ptr(row)
+    assert_in_range(stack_ptr, 0, MAX_STACK_PTR)
+
+    # 3. stack_ptr only increases by 0 or 1
+    if row.tag() == row_prev.tag() and get_call_id(row) == get_call_id(row_prev):
+        stack_ptr_diff = get_stack_ptr(row) - get_stack_ptr(row_prev)
+        assert_in_range(stack_ptr_diff, 0, 1)
+
+
+@is_circuit_code
+def check_storage(row: Row, row_prev: Row):
+    get_addr = lambda row: row.keys[2]
+    get_storage_key = lambda row: row.keys[4]
+
+    # TODO: cold VS warm
+    # TODO: connection to MPT on first and last access for each (address, key)
+
+    # 0. Unused keys are 0
+    assert row.keys[1] == 0
+    assert row.keys[3] == 0
+
+    # 1. First access for a set of all keys
+    #
+    # We add an extra write to set the value of the state in previous block, with rwc=0.
+    #
+    # When the set of all keys changes (first access of storage (address, key))
+    # - It must be a WRITE
+    if not all_keys_eq(row, row_prev):
+        assert row.is_write == 1 and row.rw_counter == 0
```

Comment:
This is still to be finished, right? Like adding constraints for previous and current value (to enable revert).

Reply:
@han0110 mentioned that we can remove the value_prev column.  See here https://hackmd.io/VCGteyZEQFKUjHA_A72f1Q for the current documentation of the state circuit design and here https://github.com/appliedzkp/zkevm-specs/pull/94#discussion_r788627623 for a relevant discussion.  So in this PR there's a single value column.

In the spec from this PR, there's a constraint for all READ rows which don't change keys, for which the value must equal the previous value; and no constraint for WRITE rows.  I think that works fine for reverts, as a revert does a WRITE and from the state circuit perspective, this write can be anything.  I think it will be the EVM circuit that enforces that the writes from a revert restore the correct value.

@miha-stopar What constraints in particular were you thinking about?

Reply:
I see, wasn't up to date. Thanks!